### PR TITLE
Add ElidePermutations pass to optimization level 3

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

This is a follow-up of #9523 that actually adds the `ElidePermutations` and the accompanying `FinalizeLayouts` passes to optimization level 3, and removes from there the `OptimizeSwapBeforeMeasure` as it's really just a special case of `ElidePermutations` (moreover, `OptimizeSwapBeforeMeasure` does **not** update the "final_layout" and breaks unitary equivalence between the original and the transpiled circuits).

### Details and comments

Should be rebased on top of #9523 (and in fact, separated from #9523 into a separate PR).

### CODE REVIEW

File: qiskit/quantum_info/operators/operator.py [hunk 1]
```diff
@@ -403,27 +405,38 @@ def from_circuit(
                 initial_layout=layout,
                 input_qubit_mapping={qubit: index for index, qubit in enumerate(circuit.qubits)},
             )
+
+        initial_layout = layout.initial_layout if layout is not None else None
```
Comment (MaitreyPrajapati) [inline qiskit/quantum_info/operators/operator.py]: nit: We can rewrite this as `inital_layout = layout and layout.initial_layout`

File: qiskit/quantum_info/operators/operator.py [hunk 2]
```diff
@@ -403,27 +405,38 @@ def from_circuit(
...
-            physical_to_virtual = layout.initial_layout.get_physical_bits()
-            qargs = [
-                layout.input_qubit_mapping[physical_to_virtual[physical_bit]]
-                for physical_bit in range(len(physical_to_virtual))
-            ]
-        # Convert circuit to an instruction
-        instruction = circuit.to_instruction()
-        op._append_instruction(instruction, qargs=qargs)
-        # If final layout is set permute output indices based on layout
+        from qiskit.synthesis.permutation.permutation_utils import _inverse_pattern
```
Comment (MaitreyPrajapati) [inline qiskit/quantum_info/operators/operator.py]: kind of nit: Since we are always going to execute this line of code, shouldn't this be at the top of the function? Right after the docstring?

Comment (jakelishman) [reply qiskit/quantum_info/operators/operator.py]: `qiskit.synthesis` is a logically higher part of the package hierarchy than `qiskit.quantum_info`, but this part of `Operator` is acting above its position, so we use a scoped import to avoid import cycles.

File: qiskit/quantum_info/operators/operator.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/quantum_info/operators/operator.py]: Your way is also valid, so it's mostly just a question of preference. `is not None` checks are very explicit about what's being expected, whereas the implicit truthiness and pass-through of Python's Boolean expressions is maybe a little less well-known (and technically allows more things to pass through the false-y test without an error).

Comment (MaitreyPrajapati) [reply qiskit/quantum_info/operators/operator.py]: Thats understandable. Fine with keeping it this way for code readability 👍

File: qiskit/quantum_info/operators/operator.py (hunk 2 above)
Comment (MaitreyPrajapati) [reply qiskit/quantum_info/operators/operator.py]: Oh I see. But my concern is more towards moving this to top of the function `from_circuit` and not the top of the file (which is where I think it could run into the circular import error). I might be wrong, can you please clarify?

File: qiskit/transpiler/preset_passmanagers/common.py [hunk 3]
```diff
@@ -407,6 +408,13 @@ def _direction_condition(property_set):
     return pre_opt
 
 
+def generate_post_op_passmanager():
+    """Generates the post-optimization pass manager."""
+    out = PassManager()
+    out.append(FinalizeLayouts())
+    return out
```
Comment (jakelishman) [inline qiskit/transpiler/preset_passmanagers/common.py]: Highly unimportant, but just to point out that this can also be spelled `return PassManager([FinalizeLayouts()])`.

(No need to change it, just commenting in case it's shorter for you in the future.)

File: qiskit/transpiler/preset_passmanagers/level3.py [hunk 4]
```diff
@@ -107,6 +107,8 @@ def level_3_pass_manager(pass_manager_config: PassManagerConfig) -> StagedPassMa
         "scheduling", scheduling_method, pass_manager_config, optimization_level=3
     )
 
+    post_optimization = common.generate_post_op_passmanager()
+
```
Comment (jakelishman) [inline qiskit/transpiler/preset_passmanagers/level3.py]: I think we ought to include the layout finalisation in every optimisation level, even if by our presets it won't have any work to do.  It just helps catch the case of somebody using (say) an O1 pass manager but injecting `ElideSwaps` into pre-op, which is a totally valid use of `StagedPassManager`.

Since we don't have `final_permutation` and its fixing of the update responsibilities to be local to the modifying pass, I think we need to treat `FinalizeLayouts` as a required part of the transpiler pipeline at all optimisation levels, or shift the responsibility of `FinalizeLayouts` into `PassManager._passmanager_backend` (https://github.com/Qiskit/qiskit/pull/9523#issuecomment-2089363405).

**Review by jakelishman — COMMENTED** [review]
If I understand the ordering correctly, the actual meat of this PR is commit 7a90b70cc2b7ec6c60f258341c755b5602f22d43.  Assuming that's true, this mostly looks fine to me.

**Comment by alexanderivrii** [comment]
We don't need this PR anymore, should I close it?

**Comment by jakelishman** [comment]
#9523 didn't add the pass to O3, so that part is still something we'd want to do separately.  It's just that I think `FinalizeLayouts` is no longer needed because it's built-in to our `PassManager`.

File: qiskit/transpiler/preset_passmanagers/level3.py (hunk 4 above)
Comment (jakelishman) [reply qiskit/transpiler/preset_passmanagers/level3.py]: Since #9523 moved the responsibility of `FinalizeLayouts` into `PassManager._passmanager_backend`, now we don't need the post-op stage at all, rather than needing it everywhere.

Comment (alexanderivrii) [reply qiskit/transpiler/preset_passmanagers/level3.py]: yes, this is much-much cleaner than having a separate `FinalizeLayouts` pass.

**Comment by alexanderivrii** [comment]
Ok, so the only thing that needs to be done (once ElidePermutations merges) is to replace OptimizeSwapsBeforeMeasure by ElidePermutations for level=3.

File: releasenotes/notes/add-elide-permutations-to-pipeline-077dad03bd55ab9c.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+features:
+  - |
+    The transpiler pass :class:`~.ElidePermutations`
+    runs by default with optimization level 3. Intuitively, removing
```
Comment (mtreinish) [inline releasenotes/notes/add-elide-permutations-to-pipeline-077dad03bd55ab9c.yaml]: 2 and 3 share the optimizations run during `init` now (since #12149 )

```suggestion
    runs by default with optimization level 2 and 3. Intuitively, removing
```

**Review by mtreinish — DISMISSED** [review]

File: test/python/compiler/test_transpiler.py [hunk 6]
```diff
@@ -1760,8 +1763,13 @@ def test_optimize_ecr_basis(self):
...
         res = transpile(circuit, basis_gates=["u", "ecr"], optimization_level=3, seed_transpiler=42)
-        self.assertEqual(res.count_ops()["ecr"], 1)
-        self.assertTrue(Operator(res).equiv(circuit))
+
+        # an iswap gate is equivalent to (swap, CZ) up to single-qubit rotations. Normally, the swap gate
+        # in the circuit would cancel with the swap gate of the (swap, CZ), leaving a single CZ gate that
+        # can be realized via one ECR gate. However, with the introduction of ElideSwap, the swap gate
+        # cancellation can not occur anymore, thus requiring two ECR gates for the iswap gate.
+        self.assertEqual(res.count_ops()["ecr"], 2)
+        self.assertEqual(Operator(circuit), Operator.from_circuit(res))
```
Comment (mtreinish) [inline test/python/compiler/test_transpiler.py:1772]: I guess this is the edge case of the elide permutations pass. It is a fun edge case and is unlikely to come up in practice I think.

File: test/python/transpiler/test_elide_permutations.py [hunk 7]
```diff
@@ -312,7 +313,7 @@ def test_not_run_after_layout(self):
         qc.h(1)
 
         spm = generate_preset_pass_manager(
-            optimization_level=3, initial_layout=list(range(2, -1, -1)), seed_transpiler=42
+            optimization_level=1, initial_layout=list(range(2, -1, -1)), seed_transpiler=42
         )
         spm.layout += ElidePermutations()
```
Comment (alexanderivrii) [inline test/python/transpiler/test_elide_permutations.py:318]: This was the other new failing test. Now that the `ElidePermutations` already runs in level=3, it is able to remove those SWAP gates before getting to the extra pass in line 318.

**Review by mtreinish — APPROVED** [review]
